### PR TITLE
Accessibility: Warn if multiple H1 headings used

### LINKED ISSUE #4230: Table of Contents should warn about multiple H1
The post title may or may not be a H1 for the reasons explained in #4191. Ideally, there should be a way to detect what a theme is using for the post title, or the theme should declare it... hard to do, I now.

Regardless of the post title, it is possible to add one or multiple H1 headings in the post content. The table of contents should warn users that having multiple H1 in a page is a very bad practice and educate them to not do that.

screen shot 2018-01-02 at 13 09 45

### PR DESCRIPTION
**Pull request by youknowriad**
closes #4230

This PR updates the document outline to mark `h1`s as invalid if we use more than one `h1`s.

**Testing instructions**

 - Add two `h1` headings to the post content
 - Open the document outline
 - Notice the headings are marked invalids with a message.

### CODE REVIEW

File: editor/components/document-outline/index.js [hunk 1]
```diff
@@ -25,6 +25,10 @@ const incorrectLevelContent = [
 	<br key="incorrect-break" />,
 	<em key="incorrect-message">{ __( '(Incorrect heading level)' ) }</em>,
 ];
+const multipleH1Headings = [
+	<br key="incorrect-break-h1" />,
+	<em key="incorrect-message-h1">{ __( '(Multiple H1 headings)' ) }</em>,
```
Comment (youknowriad) [inline editor/components/document-outline/index.js]: Not sure about the message here. Open to suggestions.

File: editor/components/document-outline/index.js [hunk 2]
```diff
@@ -25,6 +25,10 @@ const incorrectLevelContent = [
 	<br key="incorrect-break" />,
 	<em key="incorrect-message">{ __( '(Incorrect heading level)' ) }</em>,
 ];
+const multipleH1Headings = [
+	<br key="incorrect-break-h1" />,
```
Comment (mtias) [inline editor/components/document-outline/index.js]: Where are these classes coming from?

Comment (mtias) [inline editor/components/document-outline/index.js]: We should fix these classes.

Comment (youknowriad) [reply editor/components/document-outline/index.js]: Not sure I understand?

Comment (mtias) [reply editor/components/document-outline/index.js]: oh, just that the namespace should be the component not an arbitrary class. This predates this PR, though.

Comment (youknowriad) [reply editor/components/document-outline/index.js]: this is just the React "key" though not a classnamee :)

Comment (mtias) [reply editor/components/document-outline/index.js]: Haha, welcome 2018.

**Comment by afercia** [comment]
So basically, when merged, it will look like this:

screen shot 2018-01-03 at 17 23 37

Couple issues:
- we don't know how the post title is rendered in the frontend: it may or may not be a H1
- space is limited and we'd need a better message than `Multiple H1 headings`

Would it be possible to:
- when there's a post title and _one_ H1, use a message: `Your theme may already use a H1 for the post title`
- when there are multiple H1s, use a message: `Multiple H1 headings are not recommended`
- consider that in other languages even the current strings will probably be longer than the one in English; when a heading is nested by multiple levels, there's really a very limited space (see below). Maybe consider to make the whole dropdown larger?

screen shot 2018-01-03 at 17 36 46

Note: the headings themselves could be way longer.

File: editor/components/document-outline/index.js (hunk 1 above)
Comment (moorscode) [reply editor/components/document-outline/index.js]: Seems to be missing some actionability, why would multiple H1 headings be a bad thing?

Taking inspiration from one of the posts on Yoast.com: `H1 is the title of the page or the blog post, and you should only use it once.` - See https://yoast.com/seo-basics-how-to-use-headings-on-your-site/

Comment (afercia) [reply editor/components/document-outline/index.js]: @moorscode not sure I fully understand 🙂 
> why would multiple H1 headings be a bad thing?
> and you should only use it once

This PR tries to enforce that, warning users if they use it more than once.

File: editor/components/document-outline/index.js (hunk 2 above)
Comment (aduth) [reply editor/components/document-outline/index.js]: If we export `React.Fragment` from `element/index.js`, we could write this as:

```jsx
const multipleH1Headings = (
	<Fragment>
		<br />
		<em>{ __( '(Multiple H1 headings)' ) }</em>
	</Fragment>
);
```

(Not sure if others feel this is more or less readable)

**Comment by youknowriad** [comment]
Updated with the suggestions in https://github.com/WordPress/gutenberg/pull/4238#issuecomment-355060688

**Comment by mtias** [comment]
@afercia @youknowriad this will be much better once the title is just a block, but in the meantime we could also let themes inform us which heading level they use for post titles via:

```php
add_theme_support( 'gutenberg', array(
   'wide-images' => true,
   'post-title-heading' => 'h1',
) );
```

If this is present, we use the provided value; if it's not, we use the current "title" label. What do you think?

**Comment by afercia** [comment]
@mtias could that help also future support for custom post types that don't use the post title? For example, with a value `none`. In that case, it could be also required.

**Comment by youknowriad** [comment]
@mtias What about themes with multiple templates and CPTs, some using h1, and some not. I think this doesn't bring enough value compared to the complexity to handle all the use-cases.

File: editor/components/document-outline/test/index.js [hunk 3]
```diff
@@ -60,5 +64,12 @@ describe( 'DocumentOutline', () => {
 
 			expect( wrapper.find( 'TableOfContentsItem' ) ).toHaveLength( 2 );
 		} );
+
+		it( 'should render warnings for multiple h1 heaedings', () => {
```
Comment (moorscode) [inline editor/components/document-outline/test/index.js]: Detected typo: `heaedings`

**Comment by mtias** [comment]
Mmm, yes, probably not worth the headaches involved in all the possible variations at the template level.

**Comment by youknowriad** [comment]
All concerns seems addressed, merging

**Pull request merged**